### PR TITLE
feat: sealing: Use bitfields to manage sector numbers

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Needed for SaaS

## Proposed Changes
* Move sector number assigning logic to use stored bitfields instead of stored counters
* Make it possible to reserve ranges of sector numbers
  * Using `lotus-miner sectors numbers reserve/free` cli or `SectorNumReserve/Free` API.
* Add a command which makes it possible to see the sector assigner state - `lotus-miner sectors numbers info`
* Add a command which makes it possible to view sector number reservations - `lotus-miner sectors numbers reservations`

Some cli examples:
```
$ lotus-miner sectors numbers info
Next free: 4
Allocated: 0-3,11-12
Reserved: 0

$ lotus-miner sectors pledge
Created CC sector:  4

$ lotus-miner sectors numbers info
Next free: 5
Allocated: 0-4,11-12
Reserved: 0

$ lotus-miner sectors numbers reserve test-res 5-9

$ lotus-miner sectors numbers reservations
reservations: 1
test-res: count=5 5-9

$ lotus-miner sectors numbers info
Next free: 10
Allocated: 0-4,11-12
Reserved: 5-9

$ lotus-miner sectors pledge
Created CC sector:  10

## sectors 11/12 were already allocated for whatever reason, so the assigner will assign 13, which is the next available number

$ lotus-miner sectors numbers info
Next free: 13
Allocated: 0-4,10-12
Reserved: 5-9

$ lotus-miner sectors pledge
Created CC sector:  13

## if we drop the reservation without any sector numbers in that being used, the assigner will now try to reuse those numbers to reduce number gaps

$ lotus-miner sectors numbers free test-res
$ lotus-miner sectors numbers info
Next free: 5
Allocated: 0-4,10-13
Reserved: 0

$ lotus-miner sectors pledge
Created CC sector:  5
```

## TODO
* [x] Make sealing pipeline use bitfields instad of a stored sector counter for tracking sector number assignments
  * Backwards compatible with the old stored counter so upgrading/downgrading lotus-miner is free of manual migrations
* [x] Add commands/utils to allocate / reserve blocks of sector numbers
  * Makes it possible for SaaS to make sure lotus won't use specified ranges of sector numbers
* [x] Make sector number allocation logic look at miner-actor state to not create sectors with numbers which are marked as allocated
* [x] Describe how to use here
* [x] Test all this
  * [x] Itests
  * [x] Manual testing

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Comment by magik6k** [comment]
(this will get one more commit adding an additional method for creating reservations)

File: storage/pipeline/numassign.go [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+	reserved, err := m.loadBitField(ctx, reservedSectorsKey)
+	if err != nil {
+		return api.NumAssignerMeta{}, xerrors.Errorf("loading reserved sectors bitfield: %w", err)
+	}
+	allocated, err := m.loadBitField(ctx, allocatedSectorsKey)
+	if err != nil {
+		return api.NumAssignerMeta{}, xerrors.Errorf("loading allocated sectors bitfield: %w", err)
+	}
+
+	// if the allocated bitfield doesn't exist, crate it from the legacy sector counter
```
Comment (geoff-vball) [inline storage/pipeline/numassign.go:101]: typo

File: storage/pipeline/numassign.go [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+	}
+	allReserved := nm.Reserved
+
+	allReserved, err = bitfield.SubtractBitField(allReserved, cur)
+	if err != nil {
+		return xerrors.Errorf("allReserved - cur: %w", err)
+	}
+
+	// check if the reservation is colliding with any other reservation
+	coliding, err := bitfield.IntersectBitField(allReserved, reserving)
```
Comment (geoff-vball) [inline storage/pipeline/numassign.go:253]: typo in colliding?

File: cmd/lotus-miner/sectors.go [hunk 3]
```diff
@@ -2201,3 +2203,142 @@ var sectorsCompactPartitionsCmd = &cli.Command{
...
+
+var sectorsNumbersReserveCmd = &cli.Command{
+	Name:  "reserve",
+	Usage: "create sector number reservations",
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "force",
+			Usage: "skip duplicate reservation checks (note: can lead to damaging other reservations on free)",
+		},
+	},
```
Comment (geoff-vball) [inline cmd/lotus-miner/sectors.go:2303]: What's the use-case for wanting this flag?

File: lib/strle/human.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+			}
+		}
+
+		if start < last {
+			return bitfield.BitField{}, xerrors.Errorf("run %d start(%d) was less than last run end(%d)", i, start, last)
+		}
+
+		if start == last && last > 0 {
+			return bitfield.BitField{}, xerrors.Errorf("run %d start(%d) was equal to last run end(%d)", i, start, last)
+		}
```
Comment (geoff-vball) [inline lib/strle/human.go:51]: These two cases can be combined into one check of `if start <= last && last > 0`

File: cmd/lotus-miner/sectors.go (hunk 3 above)
Comment (magik6k) [reply cmd/lotus-miner/sectors.go:2303]: Changing existing reservation size / splitting reservations 'atomically', possibly some other cases too.

**Pull request merged**